src/odemis/gui/comp/overlay/world.py

```diff
@@ -1147,27 +1109,195 @@ def _points_to_rect(left, top, right, bottom):
         """ Transform two (x, y) points into a (x, y, w, h) rectangle """
         return left, top, right - left, bottom - top
 
-    def _update_view_positions(self, b_start_pos, b_end_pos):
-        """Given that the canvas has been shifted or rescaled, update the view positions of the ruler"""
-        logging.debug("Updating view position of ruler %s", self)
-        self.v_start_pos = Vec(self.cnvs.buffer_to_view(b_start_pos))
-        self.v_end_pos = Vec(self.cnvs.buffer_to_view(b_end_pos))
+    @abstractmethod
+    def sync_with_canvas(self, shift=(0, 0), scale=1.0):
+        """
+        Update the view positions of the line when the canvas has been shifted or rescaled.
+        Args:
+            shift: shift of the canvas to know whether it has changed
+            scale: scale of the canvas to know whether it has changed
+        """
+        pass
+
+    @abstractmethod
+    def draw(self, ctx, selected, canvas=None, font_size=None):
+        """
+        Draw the line to given context
+        Args:
+            ctx: cairo context to draw on
+            selected: if the line is selected, it gets highlighted and thicker
+            canvas: canvas on which the tools are drawn. In case of print-ready export a fake canvas is passed and the
+            gadget overlay draws on it.
+            font_size: fontsize is given in case of print-ready export
+        """
+        pass
+
```

Comment:
This class inherits from the abstract class GadgetToolInterface. I just deleted the above methods. ( Do you mean to mark it by using the "with_metaclass(ABCMeta, object)"??). Please check my latest version and let me know.

Reply:
It's good to remove these methods. You should *also* mark the class as "abstract", because it wouldn't work properly if you were directly creating an instance of it. So, mark with with `with_metaclass(ABCMeta, GadgetToolInterface)`.